src/controller.js

```diff
@@ -0,0 +1,601 @@
+      ctx = ctx.parent();
+      ctx = ctx.parent();
+    }
+
+    return doc.toString().toString();
+  }
+
+  /**
+   * Generate the XML body required to set tags on an object.
+   *
+   * This function tags an object which is key-value pairings
+   * that represent each tag.
+   *
+   * EXAMPLE:
+   * <Tagging>
+   *    <TagSet>
+   *       <Tag>
+   *          <Key>tag1</Key>
+   *          <Value>val1</Value>
+   *       </Tag>
+   *       <Tag>
+   *          <Key>tag2</Key>
+   *          <Value>val2</Value>
+   *       </Tag>
+   *    </TagSet>
+   * </Tagging>
+   */
+  __generateTagSetBody(tags) {
+    // TODO: Figure out how to double check that all keys and values
+    // match a pattern or something
+    tags = runSchema(tags, Joi.object().required());
+
+    let doc = new libxml.Document();
+
+    let ctx = doc.node('Tagging');
+    ctx = ctx.node('TagSet');
+    for (let key in tags) {
+      ctx = ctx.node('Tag');
+      ctx = ctx.node('Key', key);
+      ctx = ctx.parent();
+      ctx = ctx.node('Value', tags[key]);
+      ctx = ctx.parent();
+      ctx = ctx.parent();
+    }
+
+    return doc.toString().trim();
+  }
+
+  /**
+   * General method for extracting a specific property from an S3
+   * response.  This assumes it's a top level node in the main container,
+   * the response has a Bucket and Key property and those match
+   * the passed in ones.  This is a sanity check.
+   */
+  __getResponseProperty(doc, container, property, bucket, key) {
+    if (doc.root().name() !== container) {
+      throw new Error('Document does not have ' + container);
+    }
+
+    let lookingfor, foundBucket, foundKey;
+    
+    for (let child of doc.root().childNodes()) {
+      switch (child.name()) {
+        case 'Bucket':
+          foundBucket = child.text();
+          break;
+        case 'Key':
+          foundKey = child.text();
+          break;
+        case property:
+          lookingfor = child.text();
+          break;
+      }
+    }
+
+    if (foundBucket !== bucket) {
+      throw new Error('Document contains incorrect Bucket');
+    }
+
+    if (foundKey !== key) {
+      throw new Error('Document contains incorrect Key');
+    }
+
+    if (typeof lookingfor === 'undefined') {
+      throw new Error('Document does not contain ' + property);
+    }
+
+    return lookingfor;
+  }
+  /**
+   * Obtain an UploadId from an Initiate Multipart Upload response body
+   */
+  __getUploadId(doc, bucket, key) {
+    return this.__getResponseProperty(doc, 'InitiateMultipartUploadResult', 'UploadId', bucket, key);
+  }
+
+  /**
+   * Obtain the ETag for the committed multipart upload
+   */
+  __getMultipartEtag(doc, bucket, key) {
+    return this.__getResponseProperty(doc, 'CompleteMultipartUploadResult', 'ETag', bucket, key);
+  }
+
+  // Return a list of 2-tuple's that are header key and value pairings for the
+  // headers to specify as the ACL for an object
+  __determinePermissionsHeaders(permissions) {
```

Comment:
> For the record: I'm super excited about the testability of the client implementation... This is something we've neglected a bit before :)

Me too!  And using the `Controller` class in the queue means that we make sure that our ability to sign and run S3 requests is always there, both by testing this library and by integreation testing the Queue :)